src/cascadia/ShellExtension/dllmain.cpp

```diff
@@ -0,0 +1,86 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT license.
+
+#include "pch.h"
+#include "OpenTerminalHere.h"
+// NOTE: All this file is pretty egregiously taken from PowerToys's PowerRename,
+// specifically:
+//
+// https://github.com/microsoft/PowerToys/blob/master/
+//   src/modules/powerrename/dll/dllmain.cpp
+//
+// I'm not positive how much of it we need, but we definitely need:
+// * a ClassFactory that can create our implementation of IExplorerCommand
+// * a DllGetClassObject that will return the aforementioned class factory.
+
+std::atomic<DWORD> g_dwModuleRefCount = 0;
+HINSTANCE g_hInst = 0;
+
+extern "C" IMAGE_DOS_HEADER __ImageBase;
+void ModuleAddRef()
+{
+    g_dwModuleRefCount++;
+}
+
+void ModuleRelease()
+{
+    g_dwModuleRefCount--;
+}
+
+struct ShellExtClassFactory : winrt::implements<ShellExtClassFactory, IClassFactory>
+{
+public:
+    ShellExtClassFactory(_In_ REFCLSID clsid) :
+        m_clsid{ clsid } {};
+
+    // IClassFactory methods
+    IFACEMETHODIMP CreateInstance(_In_opt_ IUnknown* punkOuter,
+                                  _In_ REFIID riid,
+                                  _Outptr_ void** ppv)
+    {
+        *ppv = NULL;
+        HRESULT hr;
+        if (punkOuter)
+        {
+            hr = CLASS_E_NOAGGREGATION;
+        }
+        else if (m_clsid == __uuidof(OpenTerminalHere))
+        {
+            hr = winrt::make<OpenTerminalHere>()->QueryInterface(riid, ppv);
+        }
+        else
+        {
+            hr = CLASS_E_CLASSNOTAVAILABLE;
+        }
+        return hr;
+    }
+
+    IFACEMETHODIMP LockServer(BOOL bLock)
+    {
+        if (bLock)
+        {
+            ModuleAddRef();
+        }
+        else
+        {
+            ModuleRelease();
+        }
+        return S_OK;
+    }
+
+private:
+    CLSID m_clsid;
+};

```

Comment:
I wonder if there's a way for us to use the standard WRL Module helper to implement this. I know it's something WRL can do.

Reply:
Turns out YES!

> CoCreatableClass creates a factory for the class that can be used with functions such as CoCreateInstance.

Docs [here](https://docs.microsoft.com/en-us/cpp/cppcx/wrl/how-to-create-a-classic-com-component-using-wrl?view=vs-2019)